### ISSUE TITLE
single-namespace should not overwrite existing deployments

### ISSUE THREAD

**Issue by wmorgan** [linkerd/linkerd2#1848]
If you install Linkerd with `--single-namespace --linkerd-namespace foo`, and `foo` has any deployments that have the same name as a control plane deployment, they get silently overwritten. I think we should either check for this use case, or rename the control plane components to start with `linkerd-`.

E.g. with the emojivoto demo app, if I install Linkerd in the `emojivoto` namespace, I get (note the two `web` pods and replicasets, but only one `web` deployment):

```
william:~/devel$ kubectl get all --namespace emojivoto
NAME                              READY     STATUS    RESTARTS   AGE
pod/controller-bccb8cf76-dv5s6    4/4       Running   0          10m
pod/emoji-d9c7866bb-5g7kp         1/1       Running   0          11m
pod/grafana-b7957c8df-ck6ww       2/2       Running   0          10m
pod/prometheus-7c9966b699-trsr8   2/2       Running   0          10m
pod/vote-bot-644b8cb6b4-dbkgp     1/1       Running   0          11m
pod/voting-65b9fffd77-7rzbq       1/1       Running   0          11m
pod/web-59dd497466-488gf          2/2       Running   0          10m
pod/web-6cfbccc48-7jlwc           1/1       Running   0          11m

NAME                 TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
service/api          ClusterIP      10.99.176.154    <none>        8085/TCP            10m
service/emoji-svc    ClusterIP      None             <none>        8080/TCP            11m
service/grafana      ClusterIP      10.111.61.189    <none>        3000/TCP            10m
service/prometheus   ClusterIP      10.100.78.239    <none>        9090/TCP            10m
service/proxy-api    ClusterIP      10.104.226.77    <none>        8086/TCP            10m
service/voting-svc   ClusterIP      None             <none>        8080/TCP            11m
service/web          ClusterIP      10.101.115.227   <none>        8084/TCP,9994/TCP   10m
service/web-svc      LoadBalancer   10.107.99.240    localhost     80:30590/TCP        11m

NAME                         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/controller   1         1         1            1           10m
deployment.apps/emoji        1         1         1            1           11m
deployment.apps/grafana      1         1         1            1           10m
deployment.apps/prometheus   1         1         1            1           10m
deployment.apps/vote-bot     1         1         1            1           11m
deployment.apps/voting       1         1         1            1           11m
deployment.apps/web          1         1         1            1           11m

NAME                                    DESIRED   CURRENT   READY     AGE
replicaset.apps/controller-bccb8cf76    1         1         1         10m
replicaset.apps/emoji-d9c7866bb         1         1         1         11m
replicaset.apps/grafana-b7957c8df       1         1         1         10m
replicaset.apps/prometheus-7c9966b699   1         1         1         10m
replicaset.apps/vote-bot-644b8cb6b4     1         1         1         11m
replicaset.apps/voting-65b9fffd77       1         1         1         11m
replicaset.apps/web-59dd497466          1         1         1         10m
replicaset.apps/web-6cfbccc48           1         1         1         11m
```

**Comment by klingerf** [comment]
This came up as part of the original `--single-namespace` implementation (#1721), and I agree that the current setup is suboptimal, given that many of the controller components have fairly generic names.

It's not totally straightforward to fix it, unfortunately. Prefixing all of our names with `linkerd-` would prevent conflicts, but there's a lot of code in the CLI and controller that uses hardcoded component names. Renaming our Service objects is particularly problematic, since that changes the DNS names that we use to connect to Linkerd's public API, which means that every CLI subcommand that connects to the API would need to know if the controller was installed in single-namespace mode or not.

I haven't given it too much though, but this type of configuration feels like something that could be fixed by a CLI context file. When you configure the CLI to connect to an instance of the control plane, it could save that configuration to a local file that could be loaded automatically on subsequent invocations. But implementing contexts correctly would be a nontrivial amount of work.

**Comment by klingerf** [comment]
A simpler fix for this issue would be to prefix all of the objects in the linkerd control plane with `linkerd-`, regardless of whether or not the `--single-namespace` flag is present. That shouldn't be a prob for running in cluster-ride mode, and it would allow us to avoid a bunch of special casing for single-namespace mode. But it would be a breaking changes for all existing linkerd CLIs and control plane deployments, so we'd need to be very thoughtful about how we role it out.

**Comment by grampelberg** [comment]
It feels like an okay breaking change. Because the name changes, it is possible to operate the old and new control planes (and data planes) at the same time. The toughest bit is getting all the CLIs updated, but that doesn't feel like the end of the world.

tldr. I'd love to make this change sooner rather than later and make sure we explain the upgrade explicitly.
